### DISCUSSION TITLE
Problem with stlib module

### DISCUSSION THREAD

**Discussion by Achiga-wuhu** [SofaDefrost/SoftRobots#113] [Help - Building / Configuring] [answered]
I download SOFA of 21.06 version. The Plugins SofaPython3, Softrobots, STLIB is active in my plugin manager.
But when I operate the firststeps-tuto files, the error comes with 
`[ERROR]   [SofaPython3::SceneLoader] ModuleNotFoundError: No module named 'stlib'`
and i have already add stlib to PYTHONPATH with  \plugins\STLIB\lib\python3\site-packages\stlib3; can anyone help with this problem?
Thank you very much!

**Comment by Achiga-wuhu** [comment]
The whole error code:
```
[ERROR]   [SofaPython3::SceneLoader] ModuleNotFoundError: No module named 'stlib'

At:
  firststeps-tuto.pyscn(1): <module>
  <frozen importlib._bootstrap>(219): _call_with_frames_removed
  <frozen importlib._bootstrap_external>(728): exec_module
  <frozen importlib._bootstrap>(677): _load_unlocked
  <frozen importlib._bootstrap>(696): _load
  C:\Users\15837\AppData\Local\Programs\Python\Python37\Lib\imp.py(172): load_source
  C:\Users\15837\AppData\Local\Programs\Python\Python37\Lib\imp.py(235): load_module
  <string>(3): <module>
```

**Comment by EulalieCoevoet** [comment]
Hi, 

I see you're using SofaPython3. Can you give us the path of the scene you're trying to run?
For the moment we want to maintain the examples for both python2 and python3 users. So if you're using the SofaPython3 plugin, you should only run the examples/tutorials under repositories named `sofapython3`. My guess is that it is not the case here, because of the import of stilb (it should be stlib3 for python3). 

You can also see my comment [here](https://github.com/SofaDefrost/SoftRobots/discussions/98#discussioncomment-2229350).

**Reply by Achiga-wuhu** [reply]
Thank you very much. I operate the scene named firststeps.py3scn in 

> SOFA_v21.06.03_Win64\plugins\SoftRobots\docs\tutorials\FirstSteps\details\sofapython3.

And I add code from the firststep tutorial to the root file with 
```
import SofaRuntime
from stlib.scene import MainHeader, ContactHeader
from stlib.scene.scene import Scene
from stlib.physics.rigid import Floor, Cube
def createScene(rootNode):
    """This is my first scene"""
    SofaRuntime.importPlugin('SofaRigid')
    SofaRuntime.importPlugin('SofaConstraint')
    rootNode.addObject("OglGrid", nbSubdiv=10, size=1000)
    MainHeader(rootNode, gravity=[0.0,-981.0,0.0])
    ContactHeader(rootNode, alarmDistance=15, contactDistance=10)
    rootNode.VisualStyle.displayFlags = "showCollisionModels"
    Floor(rootNode, translation=[0.0,-160.0,0.0],  isAStaticObject=True)
    cube = Cube(rootNode, translation=[0.0,0.0,0.0], uniformScale=20.0)
    cube.addObject('UncoupledConstraintCorrection')
    return rootNode
```
And I already set the PYTHONPATH with 
> C:\SOFA_v21.06.03_Win64\SOFA_v21.06.03_Win64\plugins\STLIB\lib\python3\site-packages\stlib3
> C:\SOFA_v21.06.03_Win64\SOFA_v21.06.03_Win64\plugins\SofaPython3\lib\python3\site-packages
> C:\SOFA_v21.06.03_Win64\SOFA_v21.06.03_Win64\plugins\STLIB\lib\python3\site-packages\splib3

but after I saved the root file and reload the scene, I got this error message:

```
[ERROR]   [SofaPython3::SceneLoader] ModuleNotFoundError: No module named 'stlib'

At:
  firststeps.py3scn(3): <module>
  <frozen importlib._bootstrap>(219): _call_with_frames_removed
  <frozen importlib._bootstrap_external>(728): exec_module
  <frozen importlib._bootstrap>(677): _load_unlocked
  <frozen importlib._bootstrap>(696): _load
  C:\Users\15837\AppData\Local\Programs\Python\Python37\Lib\imp.py(172): load_source
  C:\Users\15837\AppData\Local\Programs\Python\Python37\Lib\imp.py(235): load_module
  <string>(3): <module>
```
Is there anything wrong with the PYTHONPATH I set?

**Reply by EulalieCoevoet** [reply]
I don't think there is something wrong with the PYTHONPATH. As I said the module `stlib` is for python2. Since you're using python3 you should import `stlib3`.

I'm not sure what you mean by "I add code from the firststep tutorial to the root file with...", but you can maybe checkout the files [here](https://github.com/SofaDefrost/SoftRobots/tree/master/docs/sofapython3/tutorials/FirstSteps).

**Reply by Achiga-wuhu** [reply]
Thank you so much. But after I tried another CableGripper tutorial in sofapython3 folder, the ModuleNotFoundError still occured. 
```
[ERROR]   [SofaPython3::SceneLoader] ModuleNotFoundError: No module named 'stlib3'

At:
  step0.py3scn(2): <module>
  <frozen importlib._bootstrap>(219): _call_with_frames_removed
  <frozen importlib._bootstrap_external>(728): exec_module
  <frozen importlib._bootstrap>(677): _load_unlocked
  <frozen importlib._bootstrap>(696): _load
  C:\Users\15837\AppData\Local\Programs\Python\Python37\Lib\imp.py(172): load_source
  C:\Users\15837\AppData\Local\Programs\Python\Python37\Lib\imp.py(235): load_module
  <string>(3): <module>
```
I may try linux version of SOFA because it always crashes on my Win11 system. Will it be more stable on Ubuntu VM？

**Reply by EulalieCoevoet** [reply]
Hmm okay, so now there is something wrong with the PYTHONPATH ^^

I think with Windows it is important to add the backslash at the end of your path. So try with:

```
C:\SOFA_v21.06.03_Win64\SOFA_v21.06.03_Win64\plugins\STLIB\lib\python3\site-packages\stlib3\
C:\SOFA_v21.06.03_Win64\SOFA_v21.06.03_Win64\plugins\SofaPython3\lib\python3\site-packages\
C:\SOFA_v21.06.03_Win64\SOFA_v21.06.03_Win64\plugins\STLIB\lib\python3\site-packages\splib3\
```

If it still doesn't work. Can you run python3 in a terminal and display the path:

```bash
python3
>>> import sys
>>> print(sys.path)
```

and check that the path to stlib3 is there. If it's not there, can you tell me how you've set the PYTHONPATH ? If it is there, can you try to `import stlib3` and tell me if you get an import error.

**Reply by EulalieCoevoet** [reply]
@Achiga-wuhu any update?

**Comment by andersonnardin** [comment, marked as answer]
Hello, I am facing a similar problem. I tried with SOFA 21.06.03 and 21.12.00 together with Python 2.7.18 and Python 3.7.9.
I added via Plugin manager the plugins STLIB, SoftRobots, and Python3 (without it the tutorial first-steps scenes cannot be opened).
If I try to open step1.py3scn I get:

`[ERROR]   [SofaPython3::SceneLoader] Unable to completely load the scene from file 'C:/Users/ander/Downloads/SOFA/SOFA_v21.12.00_Win64/plugins/SoftRobots_v21.12_for-SOFA-v21.12_Windows/docs/tutorials/FirstSteps/details/sofapython3/step1.py3scn'.
Python exception:
  ModuleNotFoundError: No module named 'stlib3'`

or if I try to open step1.pyscn:

`[ERROR]   [SofaPython3::SceneLoader] Unable to completely load the scene from file 'C:/Users/ander/Downloads/SOFA/SOFA_v21.12.00_Win64/plugins/SoftRobots_v21.12_for-SOFA-v21.12_Windows/docs/tutorials/FirstSteps/details/step1.pyscn'.
Python exception:
  ModuleNotFoundError: No module named 'stlib'`

I tried to add the paths to the system environment, but the problem persists. Can someone help me to open some of these files?

**Reply by andersonnardin** [reply]
Ok, guys, I think I found some workarounds to deal with the problems.
In version 21.06.03, I have to remove and add the STLIB dll file via Plugin Manager every time I close and reopen SOFA.
Or adding the following code to the scene (because PYTHONPATH is not working):

`import sys
sys.path.append("C:\\Users\\...\\SOFA_v21.12.00_Win64\\plugins\\STLIB\\lib\\python3\\site-packages")`

In version 21.12.00, the STLIB plugin provided here: https://github.com/SofaDefrost/STLIB/releases/tag/release-v21.12 is missing the folder python3 (...plugins\STLIB\lib\python3). Solution: I copied the folder from version 21.06.03. Then it works if I remove STLIB and add it again every time I run runSofa.exe (or using the code above).

Please let me know whether this is a bug (maybe a misplacing in the version management) or it should actually work this way.
Thank you for your support and hard work!

**Reply by Achiga-wuhu** [reply]
Thank you so much. I also have solved this problem by removing and adding STLB dll file in Plugin Manager every time I reopen SOFA.  The correct location for STLIB.dll should be C: ...\SOFA_v21.06.03_Win64\plugins\STLIB\bin\STLIB.dll.

**Comment by damienmarchal** [comment]
Hi all,

Thank you for reporting that problems and providing solutions. 
We opened an issue (https://github.com/SofaDefrost/STLIB/issues/72) to have it sorted out for the next release (the 22.06).
